integration_tests/src/main/python/data_gen.py

```diff
@@ -27,12 +27,43 @@
 from conftest import skip_unless_precommit_tests
 import time
 import os
+from functools import lru_cache
+import hashlib
 
 # set time zone to UTC for timestamp test cases to avoid `datetime` out-of-range error:
 # refer to: https://github.com/NVIDIA/spark-rapids/issues/7535
 os.environ['TZ'] = 'UTC'
 time.tzset()
 
+# hash object with nested data types for caching
+def hash_object(obj):
+    h = hashlib.blake2b(digest_size=10)
+    if isinstance(obj, (int, float, str, bool, type(None))):
+        h.update(str(obj).encode())
```

Comment:
It might be nice to add a comment on why `str(obj)` is needed. 

I half expected `hash_object` to be a call to an API provided by python or another package and I am curious if `pickle.dumps` accomplishes part of it (not for dictionaries, we'd have to make sure to convert to an ordered dictionary).

Reply:
Thanks for review! `update` needs bytes-like object so I convert them to string, then encode to utf-8. Added a comment there.
I think `pickle.dumps` can serialize the DataGen so we can hash on it, but it may need to override the `__reduce__` method to handle things like internal states, lambdas, and datagens nested in dictionaries, so it won't make the solution easier. I guess we need to implement the `hash_object` because it's basically an override to `__hash__` for our custom mutable objects.

Reply:
Also, for `str(obj)` I think part of it is that if these basic types fellthrough to the else:
```
h.update(str(hash(obj)).encode())
```
We could end up with undesired results. It looks like `hash(True)==hash(1)` for example, but `hash(str(True))!=hash(str(1))`. But that said, I think in this case, it might be unlikely we'll find a collision since we are traversing a set number of arguments and we follow a pattern.

Reply:
Yes, but in terms of hashing here for caching purposes, you may want to encode the type somehow...

Reply:
Should we just force each DataGen instance to have a method that returns something that can be a part of the hash, and then it is up to the DataGen to do this properly?  We are trying to make something that is generic for everything, but I don't think that is going to work or else the hashing library would have done it already.

As a side note can use look at simple equality instead? The speedup is mostly going to come from run on CPU then run the same thing on the GPU.  We might not need to solve this generically for everything if we can just check that the two instances are the same, no hashing needed.

Reply:
> Yes, but in terms of hashing here for caching purposes, you may want to encode the type somehow...

lru_cache has a parameter `typed` which is `False` by default for some reason. But we can activate it. https://github.com/python/cpython/blob/60f8117d0c685c2923b7cb17b725b67cd41e8410/Lib/functools.py#L485

Reply:
Thanks all! I also feel that the `hash_object` may not work as excepted at some edge cases and it is not easy to debug, but I didn't figured out a better solution then. 

Now I followed @revans2 's advice above to use `__repr__` to check the equality instead. 

My only concerns about this solution is it requires developers to implement `__repr__` of their custom datagens properly, so that it contains all the information that differs form others, otherwise there may be a collision that two different datagens are treated as the same by the cache.